test/Microsoft.ML.Tests/Transformers/CustomTransformerTests.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Core.Data;
+using Microsoft.ML.Runtime.Api;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Data.IO;
+using Microsoft.ML.Runtime.Model;
+using Microsoft.ML.Runtime.RunTests;
+using Microsoft.ML.Runtime.Tools;
+using System.ComponentModel.Composition;
+using System.ComponentModel.Composition.Hosting;
+using System.IO;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.Tests.Transformers
+{
+    public sealed class CustomTransformerTests : TestDataPipeBase
+    {
+        public CustomTransformerTests(ITestOutputHelper helper) : base(helper)
+        {
+        }
+
+        public class MyInput
+        {
+            public float Float1 { get; set; }
+            public float[] Float4 { get; set; }
+        }
+
+        public class MyOutput
+        {
+            public string Together { get; set; }
+        }
+
+        public class MyLambda
+        {
+            [Export("MyLambda")]
+            public ITransformer MyTransformer => new CustomMappingTransformer<MyInput, MyOutput>(ML, MyAction, "MyLambda");
```

Comment:
>"MyLambda" [](start = 111, length = 10)

So the user has to specify `"MyLambda"` a couple places. Once in the attribute, once in the transformer constructor (here where I put the comment), and once again when performing this mapping. It seems like this *could* be discoverable by reflection, so as to avoid the redundancy?

Reply:
The attribute is not the only way to inject components (it's just the shortest and the most documented on the web. So, you cannot rely on the attribute as the source of the contract name.

Also note that the codebase that does the training and codebase that does the scoring don't have to be the same. Only the scoring one needs to inject the transformer. It's easier to pass just `string contractName` across this potential code boundary than any other information.

---
In reply to: [229428417](https://github.com/dotnet/machinelearning/pull/1406#discussion_r229428417) [](ancestors = 229428417)

Reply:
Actually, we could put `null` here instead of `"MyLambda"`. The test would continue to work, but loaded model will stop being save-able.

---
In reply to: [229462099](https://github.com/dotnet/machinelearning/pull/1406#discussion_r229462099) [](ancestors = 229462099,229428417)